<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Annotated gene list | Extending iSEE</title>
  <meta name="description" content="This book describes how to use the Bioconductor iSEE package to create web-applications for exploring data stored in SummarizedExperiment objects." />
  <meta name="generator" content="bookdown 0.17 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Annotated gene list | Extending iSEE" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book describes how to use the Bioconductor iSEE package to create web-applications for exploring data stored in SummarizedExperiment objects." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Annotated gene list | Extending iSEE" />
  
  <meta name="twitter:description" content="This book describes how to use the Bioconductor iSEE package to create web-applications for exploring data stored in SummarizedExperiment objects." />
  

<meta name="author" content="Kevin Rue-Albrecht, Federico Marini, Charlotte Soneson, and Aaron Lun" />


<meta name="date" content="2020-02-23" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="dynamic-differential-expression.html"/>
<link rel="next" href="gene-ontology-table.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Extending <i>iSEE</i></a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="panels.html"><a href="panels.html"><i class="fa fa-check"></i><b>1</b> Panel classes</a><ul>
<li class="chapter" data-level="1.1" data-path="panels.html"><a href="panels.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="panels.html"><a href="panels.html#the-panel-class"><i class="fa fa-check"></i><b>1.2</b> The Panel class</a></li>
<li class="chapter" data-level="1.3" data-path="panels.html"><a href="panels.html#the-dotplot-and-table-panel-families"><i class="fa fa-check"></i><b>1.3</b> The DotPlot and Table panel families</a></li>
<li class="chapter" data-level="1.4" data-path="panels.html"><a href="panels.html#the-columndotplot-and-rowdotplot-panel-families"><i class="fa fa-check"></i><b>1.4</b> The ColumnDotPlot and RowDotPlot panel families</a></li>
<li class="chapter" data-level="1.5" data-path="panels.html"><a href="panels.html#built-in-columndotplot-panel-classes"><i class="fa fa-check"></i><b>1.5</b> Built-in ColumnDotPlot panel classes</a></li>
<li class="chapter" data-level="1.6" data-path="panels.html"><a href="panels.html#built-in-rowdotplot-panel-classes"><i class="fa fa-check"></i><b>1.6</b> Built-in RowDotPlot panel classes</a></li>
<li class="chapter" data-level="1.7" data-path="panels.html"><a href="panels.html#the-columntable-and-rowtable-panel-families"><i class="fa fa-check"></i><b>1.7</b> The ColumnTable and RowTable panel families</a></li>
<li class="chapter" data-level="1.8" data-path="panels.html"><a href="panels.html#built-in-columntable-panel-classes"><i class="fa fa-check"></i><b>1.8</b> Built-in ColumnTable panel classes</a></li>
<li class="chapter" data-level="1.9" data-path="panels.html"><a href="panels.html#built-in-rowtable-panel-classes"><i class="fa fa-check"></i><b>1.9</b> Built-in RowTable panel classes</a></li>
<li class="chapter" data-level="1.10" data-path="panels.html"><a href="panels.html#complexheatmapplot-class"><i class="fa fa-check"></i><b>1.10</b> The ComplexHeatmapPlot panel class</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="api.html"><a href="api.html"><i class="fa fa-check"></i><b>2</b> The panel API</a><ul>
<li class="chapter" data-level="2.1" data-path="api.html"><a href="api.html#cachecommoninfo"><i class="fa fa-check"></i><b>2.1</b> .cacheCommonInfo</a></li>
<li class="chapter" data-level="2.2" data-path="api.html"><a href="api.html#refineparameters"><i class="fa fa-check"></i><b>2.2</b> .refineParameters</a></li>
<li class="chapter" data-level="2.3" data-path="api.html"><a href="api.html#generatedotplot"><i class="fa fa-check"></i><b>2.3</b> .generateDotPlot</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="server.html"><a href="server.html"><i class="fa fa-check"></i><b>3</b> The app server</a><ul>
<li class="chapter" data-level="3.1" data-path="server.html"><a href="server.html#robjects"><i class="fa fa-check"></i><b>3.1</b> Reactive objects</a></li>
<li class="chapter" data-level="3.2" data-path="server.html"><a href="server.html#persistent-non-reactive-objects"><i class="fa fa-check"></i><b>3.2</b> Persistent (non-reactive) objects</a></li>
<li class="chapter" data-level="3.3" data-path="server.html"><a href="server.html#memory"><i class="fa fa-check"></i><b>3.3</b> The app memory</a></li>
<li class="chapter" data-level="3.4" data-path="server.html"><a href="server.html#initialization-of-the-app-server"><i class="fa fa-check"></i><b>3.4</b> Initialization of the app server</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="developing.html"><a href="developing.html"><i class="fa fa-check"></i><b>4</b> Developing new panels</a><ul>
<li class="chapter" data-level="4.1" data-path="developing.html"><a href="developing.html#create-a-new-s4-class"><i class="fa fa-check"></i><b>4.1</b> Create a new S4 class</a></li>
<li class="chapter" data-level="4.2" data-path="developing.html"><a href="developing.html#add-a-constructor-function"><i class="fa fa-check"></i><b>4.2</b> Add a constructor function</a></li>
<li class="chapter" data-level="4.3" data-path="developing.html"><a href="developing.html#set-the-panel-name-in-the-gui"><i class="fa fa-check"></i><b>4.3</b> Set the panel name in the GUI</a></li>
<li class="chapter" data-level="4.4" data-path="developing.html"><a href="developing.html#define-the-commands-generating-a-plot-output"><i class="fa fa-check"></i><b>4.4</b> Define the commands generating a plot output</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html"><i class="fa fa-check"></i><b>5</b> Dynamic reduced dimensions</a><ul>
<li class="chapter" data-level="5.1" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html#overview-1"><i class="fa fa-check"></i><b>5.1</b> Overview</a></li>
<li class="chapter" data-level="5.2" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html#class-basics"><i class="fa fa-check"></i><b>5.2</b> Class basics</a></li>
<li class="chapter" data-level="5.3" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html#setting-up-the-interface"><i class="fa fa-check"></i><b>5.3</b> Setting up the interface</a></li>
<li class="chapter" data-level="5.4" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html#creating-the-observers"><i class="fa fa-check"></i><b>5.4</b> Creating the observers</a></li>
<li class="chapter" data-level="5.5" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html#making-the-plot"><i class="fa fa-check"></i><b>5.5</b> Making the plot</a></li>
<li class="chapter" data-level="5.6" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html#finishing-touches"><i class="fa fa-check"></i><b>5.6</b> Finishing touches</a></li>
<li class="chapter" data-level="5.7" data-path="dynamic-reduced-dimensions.html"><a href="dynamic-reduced-dimensions.html#in-action"><i class="fa fa-check"></i><b>5.7</b> In action</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html"><i class="fa fa-check"></i><b>6</b> Dynamic differential expression</a><ul>
<li class="chapter" data-level="6.1" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html#overview-2"><i class="fa fa-check"></i><b>6.1</b> Overview</a></li>
<li class="chapter" data-level="6.2" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html#class-basics-1"><i class="fa fa-check"></i><b>6.2</b> Class basics</a></li>
<li class="chapter" data-level="6.3" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html#setting-up-the-interface-1"><i class="fa fa-check"></i><b>6.3</b> Setting up the interface</a></li>
<li class="chapter" data-level="6.4" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html#creating-the-observers-1"><i class="fa fa-check"></i><b>6.4</b> Creating the observers</a></li>
<li class="chapter" data-level="6.5" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html#making-the-table"><i class="fa fa-check"></i><b>6.5</b> Making the table</a></li>
<li class="chapter" data-level="6.6" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html#finishing-touches-1"><i class="fa fa-check"></i><b>6.6</b> Finishing touches</a></li>
<li class="chapter" data-level="6.7" data-path="dynamic-differential-expression.html"><a href="dynamic-differential-expression.html#in-action-1"><i class="fa fa-check"></i><b>6.7</b> In action</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="annotated-gene-list.html"><a href="annotated-gene-list.html"><i class="fa fa-check"></i><b>7</b> Annotated gene list</a><ul>
<li class="chapter" data-level="7.1" data-path="annotated-gene-list.html"><a href="annotated-gene-list.html#overview-3"><i class="fa fa-check"></i><b>7.1</b> Overview</a></li>
<li class="chapter" data-level="7.2" data-path="annotated-gene-list.html"><a href="annotated-gene-list.html#class-basics-2"><i class="fa fa-check"></i><b>7.2</b> Class basics</a></li>
<li class="chapter" data-level="7.3" data-path="annotated-gene-list.html"><a href="annotated-gene-list.html#setting-up-the-interface-2"><i class="fa fa-check"></i><b>7.3</b> Setting up the interface</a></li>
<li class="chapter" data-level="7.4" data-path="annotated-gene-list.html"><a href="annotated-gene-list.html#creating-the-observers-2"><i class="fa fa-check"></i><b>7.4</b> Creating the observers</a></li>
<li class="chapter" data-level="7.5" data-path="annotated-gene-list.html"><a href="annotated-gene-list.html#in-action-2"><i class="fa fa-check"></i><b>7.5</b> In action</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html"><i class="fa fa-check"></i><b>8</b> Gene ontology table</a><ul>
<li class="chapter" data-level="8.1" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html#overview-4"><i class="fa fa-check"></i><b>8.1</b> Overview</a></li>
<li class="chapter" data-level="8.2" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html#class-basics-3"><i class="fa fa-check"></i><b>8.2</b> Class basics</a></li>
<li class="chapter" data-level="8.3" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html#setting-up-the-interface-3"><i class="fa fa-check"></i><b>8.3</b> Setting up the interface</a></li>
<li class="chapter" data-level="8.4" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html#generating-the-output"><i class="fa fa-check"></i><b>8.4</b> Generating the output</a></li>
<li class="chapter" data-level="8.5" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html#creating-the-observers-3"><i class="fa fa-check"></i><b>8.5</b> Creating the observers</a></li>
<li class="chapter" data-level="8.6" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html#handling-selections"><i class="fa fa-check"></i><b>8.6</b> Handling selections</a></li>
<li class="chapter" data-level="8.7" data-path="gene-ontology-table.html"><a href="gene-ontology-table.html#in-action-3"><i class="fa fa-check"></i><b>8.7</b> In action</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Extending <em>iSEE</em></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="annotated-gene-list" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> Annotated gene list</h1>
<div id="overview-3" class="section level2">
<h2><span class="header-section-number">7.1</span> Overview</h2>
<p>When given a gene list, we often need to look up the function of the top genes in a search engine. This typically involves copy-pasting the gene name or ID into the search box and pressing Enter, which is a pain. Instead, we can automate this process in <em><a href="https://bioconductor.org/packages/3.11/iSEE">iSEE</a></em> by creating an <strong>annotated gene table</strong>. We demonstrate by showing how we can dynamically look up annotation for each gene in the <code>rowData</code> of a <code>SummarizedExperiment</code>.</p>
</div>
<div id="class-basics-2" class="section level2">
<h2><span class="header-section-number">7.2</span> Class basics</h2>
<p>First, we define the basics of our new <code>Panel</code> class. Our new class will be showing the gene-level metadata, so we inherit from the <code>RowDataTable</code> class that does exactly this. We add some slots specifying which column of the table contains our gene IDs, the type of ID and the organism database to use.</p>
<p>We specialize the validity method to check that the <code>IDColumn</code> is either a string or <code>NULL</code>; if the latter, we assume that the ID is stored in the row name. We also add some cursory checks for the other parameters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">allowable &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ENSEMBL&quot;</span>, <span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;ENTREZID&quot;</span>)
<span class="kw">setValidity2</span>(<span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="cf">function</span>(object) {
    msg &lt;-<span class="st"> </span><span class="kw">character</span>(<span class="dv">0</span>)

    <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.null</span>(val &lt;-<span class="st"> </span>object[[<span class="st">&quot;IDColumn&quot;</span>]]) <span class="op">&amp;&amp;</span><span class="st"> </span>(<span class="kw">length</span>(val)<span class="op">!=</span>1L <span class="op">||</span><span class="st"> </span><span class="kw">is.na</span>(val))) {
        msg &lt;-<span class="st"> </span><span class="kw">c</span>(msg, <span class="st">&quot;&#39;IDColumn must be NULL or a string&quot;</span>)
    }

    <span class="cf">if</span> (<span class="op">!</span><span class="kw">isSingleString</span>(orgdb &lt;-<span class="st"> </span>object[[<span class="st">&quot;Organism&quot;</span>]])) {
        msg &lt;-<span class="st"> </span><span class="kw">c</span>(msg, <span class="kw">sprintf</span>(<span class="st">&quot;&#39;Organism&#39; should be a single string&quot;</span>, orgdb))
    }

    <span class="cf">if</span> (<span class="op">!</span><span class="kw">isSingleString</span>(type &lt;-<span class="st"> </span>object[[<span class="st">&quot;IDType&quot;</span>]]) <span class="op">||</span><span class="st"> </span><span class="op">!</span>type <span class="op">%in%</span><span class="st"> </span>allowable) {
        msg &lt;-<span class="st"> </span><span class="kw">c</span>(msg, <span class="st">&quot;&#39;IDType&#39; should be &#39;ENSEMBL&#39;, &#39;SYMBOL&#39; or &#39;ENTREZID&#39;&quot;</span>)
    }

    <span class="cf">if</span> (<span class="kw">length</span>(open &lt;-<span class="st"> </span>object[[<span class="st">&quot;AnnoBoxOpen&quot;</span>]])<span class="op">!=</span>1L <span class="op">||</span><span class="st"> </span><span class="kw">is.na</span>(open)) {
        msg &lt;-<span class="st"> </span><span class="kw">c</span>(msg, <span class="st">&quot;&#39;AnnoBoxOpen&#39; should be a non-missing logical scalar&quot;</span>)
    }

    <span class="cf">if</span> (<span class="kw">length</span>(msg)) {
        <span class="kw">return</span>(msg)
    }
    <span class="ot">TRUE</span>
})</code></pre></div>
<p>We then specialize the initialize method to set reasonable defaults for these parameters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setMethod</span>(<span class="st">&quot;initialize&quot;</span>, <span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="cf">function</span>(.Object, <span class="dt">IDColumn=</span><span class="ot">NULL</span>, 
    <span class="dt">Organism=</span><span class="st">&quot;org.Mm.eg.db&quot;</span>, <span class="dt">IDType=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">AnnoBoxOpen=</span><span class="ot">FALSE</span>, ...)
{
    <span class="kw">callNextMethod</span>(.Object, <span class="dt">IDColumn=</span>IDColumn, <span class="dt">IDType=</span>IDType,
        <span class="dt">Organism=</span>Organism, <span class="dt">AnnoBoxOpen=</span>AnnoBoxOpen, ...)
})</code></pre></div>
</div>
<div id="setting-up-the-interface-2" class="section level2">
<h2><span class="header-section-number">7.3</span> Setting up the interface</h2>
<p>We define the full name and desired default color for display purposes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setMethod</span>(<span class="st">&quot;.fullName&quot;</span>, <span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="cf">function</span>(x) <span class="st">&quot;Annotated gene table&quot;</span>)

<span class="kw">setMethod</span>(<span class="st">&quot;.panelColor&quot;</span>, <span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="cf">function</span>(x) <span class="st">&quot;#AA1122&quot;</span>)</code></pre></div>
<p>We want to add another UI element for showing the gene-level annotation. This is achieved by specializing the <code>.defineOutput()</code> method as shown below; note the prefixing by the panel name to ensure that output element IDs from different panels are unique.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setMethod</span>(<span class="st">&quot;.defineOutput&quot;</span>, <span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="cf">function</span>(x, ...) {
    panel_name &lt;-<span class="st"> </span><span class="kw">.getEncodedName</span>(x)
    <span class="kw">tagList</span>(
        <span class="kw">callNextMethod</span>(), <span class="co"># Re-using RowDataTable&#39;s definition.</span>
        <span class="kw">uiOutput</span>(<span class="kw">paste0</span>(panel_name, <span class="st">&quot;_annotation&quot;</span>)),
        <span class="kw">hr</span>()
    )
})</code></pre></div>
<p>We also set up interface elements for changing the annotation parameters. We will put these elements in a separate “Annotation parameters” collapsible box, which is initialized in an opened or closed state depending on the <code>AnnoBoxOpen</code> slot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setMethod</span>(<span class="st">&quot;.defineInterface&quot;</span>, <span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="cf">function</span>(x, se, select_info) {
    panel_name &lt;-<span class="st"> </span><span class="kw">.getEncodedName</span>(x)
    <span class="kw">c</span>(
        <span class="kw">list</span>(
            <span class="kw">collapseBox</span>(
                <span class="kw">paste0</span>(panel_name, <span class="st">&quot;_AnnoBoxOpen&quot;</span>),
                <span class="dt">title=</span><span class="st">&quot;Annotation parameters&quot;</span>,
                <span class="dt">open=</span>x[[<span class="st">&quot;AnnoBoxOpen&quot;</span>]],
                <span class="kw">selectInput</span>(<span class="kw">paste0</span>(panel_name, <span class="st">&quot;_IDColumn&quot;</span>),
                    <span class="dt">label=</span><span class="st">&quot;ID-containing column:&quot;</span>,
                    <span class="dt">choices=</span><span class="kw">colnames</span>(<span class="kw">rowData</span>(se)), 
                    <span class="dt">selected=</span>x[[<span class="st">&quot;IDColumn&quot;</span>]]
                ),
                <span class="kw">selectInput</span>(<span class="kw">paste0</span>(panel_name, <span class="st">&quot;_IDType&quot;</span>),
                    <span class="dt">label=</span><span class="st">&quot;ID type:&quot;</span>,
                    <span class="dt">choices=</span>allowable,
                    <span class="dt">selected=</span>x[[<span class="st">&quot;IDType&quot;</span>]]
                ),
                <span class="kw">selectInput</span>(<span class="kw">paste0</span>(panel_name, <span class="st">&quot;_Organism&quot;</span>),
                    <span class="dt">label=</span><span class="st">&quot;Organism&quot;</span>,
                    <span class="dt">choices=</span><span class="kw">c</span>(<span class="st">&quot;org.Hs.eg.db&quot;</span>, <span class="st">&quot;org.Mm.eg.db&quot;</span>),
                    <span class="dt">selected=</span>x[[<span class="st">&quot;Organism&quot;</span>]]
                )
            )
        ),
        <span class="kw">callNextMethod</span>()
    ) 
})</code></pre></div>
</div>
<div id="creating-the-observers-2" class="section level2">
<h2><span class="header-section-number">7.4</span> Creating the observers</h2>
<p>We specialize <code>.createObservers</code> to define some observers to respond to changes in our new interface elements. Note the use of <code>callNextMethod()</code> to ensure that observers of the parent class are also created.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setMethod</span>(<span class="st">&quot;.createObservers&quot;</span>, <span class="st">&quot;GeneAnnoTable&quot;</span>,
    <span class="cf">function</span>(x, se, input, session, pObjects, rObjects)
{
    <span class="kw">callNextMethod</span>()

    plot_name &lt;-<span class="st"> </span><span class="kw">.getEncodedName</span>(x)

    <span class="kw">.createUnprotectedParameterObservers</span>(plot_name,
        <span class="dt">fields=</span><span class="kw">c</span>(<span class="st">&quot;IDColumn&quot;</span>, <span class="st">&quot;Organism&quot;</span>, <span class="st">&quot;IDType&quot;</span>), 
        <span class="dt">input=</span>input, <span class="dt">pObjects=</span>pObjects, <span class="dt">rObjects=</span>rObjects)
})</code></pre></div>
<p>We need to set up a rendering expression for the annotation element that responds to the selected gene. By using <code>.trackSingleSelection()</code>, we ensure that this UI element updates in response to changes in the table selection. We add a series of protective measures to avoid the application crashing due to missing organism packages or unmatched IDs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setMethod</span>(<span class="st">&quot;.renderOutput&quot;</span>, <span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="cf">function</span>(x, se, ..., output, pObjects, rObjects) {
    <span class="kw">callNextMethod</span>() <span class="co"># Re-using RowDataTable&#39;s output rendering.</span>

    panel_name &lt;-<span class="st"> </span><span class="kw">.getEncodedName</span>(x)
    output[[<span class="kw">paste0</span>(panel_name, <span class="st">&quot;_annotation&quot;</span>)]] &lt;-<span class="st"> </span><span class="kw">renderUI</span>({
        <span class="kw">.trackSingleSelection</span>(panel_name, rObjects)
        instance &lt;-<span class="st"> </span>pObjects<span class="op">$</span>memory[[panel_name]]

        rowdata_col &lt;-<span class="st"> </span>instance[[<span class="st">&quot;IDColumn&quot;</span>]]
        selectedGene &lt;-<span class="st"> </span>instance[[<span class="st">&quot;Selected&quot;</span>]]
        <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.null</span>(rowdata_col)) {
            selectedGene &lt;-<span class="st"> </span><span class="kw">rowData</span>(se)[selectedGene,rowdata_col]
        }

        keytype &lt;-<span class="st"> </span>instance[[<span class="st">&quot;IDType&quot;</span>]]
        selgene_entrez &lt;-<span class="st"> </span><span class="ot">NA</span>
        <span class="cf">if</span> (keytype<span class="op">!=</span><span class="st">&quot;ENTREZID&quot;</span>) {
            ORG &lt;-<span class="st"> </span>instance[[<span class="st">&quot;Organism&quot;</span>]]
            <span class="cf">if</span> (<span class="kw">require</span>(ORG, <span class="dt">character.only=</span><span class="ot">TRUE</span>, <span class="dt">quietly=</span><span class="ot">TRUE</span>)) {
                orgdb &lt;-<span class="st"> </span><span class="kw">get</span>(ORG)
                selgene_entrez &lt;-<span class="st"> </span><span class="kw">try</span>(<span class="kw">mapIds</span>(orgdb, selectedGene, <span class="st">&quot;ENTREZID&quot;</span>, keytype), 
                    <span class="dt">silent=</span><span class="ot">TRUE</span>)
            }
        } <span class="cf">else</span> {
            selgene_entrez &lt;-<span class="st"> </span>selectedGene
        }

        <span class="cf">if</span> (<span class="kw">is.na</span>(selgene_entrez) <span class="op">||</span><span class="st"> </span><span class="kw">is</span>(selgene_entrez, <span class="st">&quot;try-error&quot;</span>)) {
            <span class="kw">return</span>(<span class="ot">NULL</span>)
        }

        fullinfo &lt;-<span class="st"> </span>rentrez<span class="op">::</span><span class="kw">entrez_summary</span>(<span class="st">&quot;gene&quot;</span>, selgene_entrez)
        link_pubmed &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;&lt;a href=&quot;http://www.ncbi.nlm.nih.gov/gene/?term=&#39;</span>,
            selgene_entrez,
            <span class="st">&#39;&quot; target=&quot;_blank&quot;&gt;Click here to see more at the NCBI database&lt;/a&gt;&#39;</span>)

        mycontent &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;&lt;b&gt;&quot;</span>,fullinfo<span class="op">$</span>name, <span class="st">&quot;&lt;/b&gt;&lt;br/&gt;&lt;br/&gt;&quot;</span>,
            fullinfo<span class="op">$</span>description,<span class="st">&quot;&lt;br/&gt;&lt;br/&gt;&quot;</span>,
            <span class="kw">ifelse</span>(fullinfo<span class="op">$</span>summary <span class="op">==</span><span class="st"> &quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="kw">paste0</span>(fullinfo<span class="op">$</span>summary, <span class="st">&quot;&lt;br/&gt;&lt;br/&gt;&quot;</span>)),
            link_pubmed)

        <span class="kw">HTML</span>(mycontent)
    })
})</code></pre></div>
<p>Observant readers will note that the body of the rendering expression uses <code>instance</code> rather than <code>x</code>. This is intentional as it ensures that we are using the parameter settings from the current state of the app. If we used <code>x</code>, we would always be using the parameters from the initial state of the app, which is not what we want.</p>
</div>
<div id="in-action-2" class="section level2">
<h2><span class="header-section-number">7.5</span> In action</h2>
<p>Let’s put our new panel to the test using the <code>sce</code> object, preprocessed in a <a href="developing.html#developing">previous chapter</a>.</p>
<p>Setting up the iSEE instance is as easy as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gat &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;GeneAnnoTable&quot;</span>, <span class="dt">PanelWidth=</span>8L)
app &lt;-<span class="st"> </span><span class="kw">iSEE</span>(sce, <span class="dt">initial=</span><span class="kw">list</span>(gat))</code></pre></div>
<p>Clicking on any row will bring up the Entrez annotation (if available) for that feature. It is probably best to click on some well-annotated genes as the set of RIKEN transcripts at the front don’t have much annotation.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dynamic-differential-expression.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="gene-ontology-table.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["iSEE-book.pdf", "iSEE-book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
